sp-sim/src/sensors.rs

```diff
@@ -0,0 +1,218 @@
+
+use crate::config::SpComponentConfig;
+use gateway_messages::measurement::MeasurementError;
+use gateway_messages::measurement::MeasurementKind;
+use gateway_messages::sp_impl::BoundsChecked;
+use gateway_messages::ComponentDetails;
+use gateway_messages::DeviceCapabilities;
+use gateway_messages::Measurement;
+use gateway_messages::SensorDataMissing;
+use gateway_messages::SensorError;
+use gateway_messages::SensorReading;
+use gateway_messages::SensorRequest;
+use gateway_messages::SensorRequestKind;
+use gateway_messages::SensorResponse;
+use gateway_messages::SpComponent;
+
+use std::collections::HashMap;
+
+pub(crate) struct Sensors {
+    by_component: HashMap<SpComponent, Vec<u32>>,
+    sensors: Vec<Sensor>,
+}
+
+#[derive(Debug)]
+struct Sensor {
+    def: SensorDef,
+    state: SensorState,
+}
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+pub struct SensorDef {
+    pub name: String,
+    pub kind: MeasurementKind,
+}
+
+// TODO(eliza): note that currently, we just hardcode these in
+// `MeasurementConfig`. Eventually, it would be neat to allow the sensor to be
+// changed dynamically as part of a simulation.
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+pub struct SensorState {
+    #[serde(default)]
+    pub last_error: Option<LastError>,
+
+    #[serde(default)]
+    pub last_data: Option<LastData>,
+}
+
+#[derive(
+    Clone, Copy, Debug, serde::Serialize, serde::Deserialize, PartialEq,
+)]
+pub struct LastError {
+    pub timestamp: u64,
+    pub value: SensorDataMissing,
+}
+
+#[derive(
+    Clone, Copy, Debug, serde::Serialize, serde::Deserialize, PartialEq,
+)]
+pub struct LastData {
+    pub timestamp: u64,
+    pub value: f32,
+}
+
+impl SensorState {
+    fn last_reading(&self) -> SensorReading {
+        match self {
+            Self { last_data: Some(data), last_error: Some(error) } => {
+                if data.timestamp >= error.timestamp {
+                    SensorReading {
+                        value: Ok(data.value),
+                        timestamp: data.timestamp,
+                    }
+                } else {
+                    SensorReading {
+                        value: Err(error.value),
+                        timestamp: error.timestamp,
+                    }
+                }
+            }
+            Self { last_data: Some(data), last_error: None } => SensorReading {
+                value: Ok(data.value),
+                timestamp: data.timestamp,
+            },
+            Self { last_data: None, last_error: Some(error) } => {
+                SensorReading {
+                    value: Err(error.value),
+                    timestamp: error.timestamp,
+                }
+            }
+            Self { last_data: None, last_error: None } => SensorReading {
+                value: Err(SensorDataMissing::DeviceNotPresent),
+                timestamp: 0, // TODO(eliza): what do?
```

Comment:
We could maybe record when `sp-sim` starts and return an actual timestamp here, or even just when the `Sensors` instance is created? I'm _not at all_ sure that's actually worth doing, unless it's useful in other contexts too (e.g., as part of the "eventually, change these dynamically over time" TODO noted above).

Reply:
The thing is that the SP's timestamp is not a wall-clock time, it's a monotonic integer counting SP clock cycles that's relative only to when the SP started. So, making up something here only makes sense if we also made up similar timestamps for other sensor measurements we return. Which I'd like to do eventually.